File: frontend/vnu-frontend/src/app/shared/types/project.ts
export interface Project {
    id: string;
    title: string;
    author: string;
    targetAudience: 'Giảng viên' | 'Sinh viên' | string;
    field: string;
    year: number;
    status: 'Đang thực hiện' | 'Đã nghiệm thu' | 'Hoàn thành' | string;
    abstract: string;
    keywords: string[];
    createdAt?: string;
    updatedAt?: string;
    ai_relevance_score?: number;
}

export interface SearchFilters {
    field?: string;
    targetAudience?: string;
    year?: number | string;
    status?: string;
    keywords?: string[];
    documentType?: string;
    implementationYear?: number | string;
}

export interface SearchResponse {
    data: {
        total: number;
        items: Project[];
    };
    status: string;
    message?: string;
}

export interface SearchError {
    code: string;
    message: string;
}

File: backend/app/models/project.py
from sqlalchemy import Column, Integer, String, Text, JSON, Computed, Index  # type: ignore
from sqlalchemy.dialects.postgresql import TSVECTOR
from sqlalchemy.orm import deferred  # type: ignore
from app.db.session import Base


class ResearchProject(Base):
    __tablename__ = "research_projects"

    id = Column(String, primary_key=True, index=True)
    title = Column(String, nullable=False)
    author = Column(String, nullable=False)
    target_audience = Column(String)  # Giáo viên, Học sinh
    field = Column(String)            # Lĩnh vực
    year = Column(Integer)
    status = Column(String)
    abstract = Column(Text)
    keywords = Column(JSON, default=[])  # Từ khóa/Tags
    document_type = Column(String, nullable=True)
    implementation_year = Column(Integer, nullable=True)

    # --- Cột chuẩn hóa cho tìm kiếm nhanh (pre-computed, indexed) ---
    search_vector = deferred(
        Column(
            TSVECTOR,
            Computed("setweight(to_tsvector('simple', unaccent(coalesce(title, ''))), 'A') || setweight(to_tsvector('simple', unaccent(coalesce(author, ''))), 'C')", persisted=True),
        )
    )

    __table_args__ = (
        Index("ix_research_projects_search_vector", "search_vector", postgresql_using="gin"),
    )
